File: lc33_SearchInRotatedArray.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 1:
            if nums[0] == target:
                return 0
            else:
                return -1
            
        if len(nums) == 2:
            if nums[0] == target:
                return 0
            elif nums[1] == target:
                return 1
            else:
                return -1

        l, r = 0, len(nums) - 1
        while l <= r:
            if l == r:
                if nums[l] != target:
                    return -1
                return l
            m = l + ((r - l) // 2)
            if nums[m] == target:
                return m
            # rememeber that this condition is checking to see if the left half of the array is sorted, not which half we should move to yet
            if nums[l] <= nums[m]:
                if nums[l] <= target and target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            else:
                if target > nums[m] and target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
        return -1
    
solution = Solution()

print(solution.search([2,3,4,5,1], 1))

File: lc33_searchInRotatedArray.java
public class lc33_searchInRotatedArray {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target)
                return mid;
            if (nums[left] <= nums[mid]) {
                if (nums[left] <= target && target < nums[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if (nums[mid] < target && target <= nums[right]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return -1;
    }

    public static void main(String args[]) {
        lc33_searchInRotatedArray solution = new lc33_searchInRotatedArray();
        int[] nums1 = { 4, 5, 6, 7, 0, 1, 2 };
        int[] nums2 = { 1 };
        System.out.printf("Case 1: %b\n", (solution.search(nums1, 0) == 4));
        System.out.printf("Case 2: %b\n", (solution.search(nums1, 3) == -1));
        System.out.printf("Case 3: %b\n", (solution.search(nums2, 0) == -1));
    }
}
